File: Kernel Simulator/TimeDate/TimeZones.vb

'    Kernel Simulator  Copyright (C) 2018-2021  EoflaOE
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.TimeZoneInfo

Public Module TimeZones

    'Time Zones in an array
    Public zones As TimeZoneInfo()
    Public zoneTimes As New Dictionary(Of String, DateTime)

    ''' <summary>
    ''' Populates current time in all of the time zones (IANA on Unix).
    ''' </summary>
    Public Sub InitTimesInZones()

        'Get all system time zones (IANA on Unix)
        zones = GetSystemTimeZones.ToArray
        Wdbg("I", "Found {0} time zones.", zones.Count)

        'Run a cleanup in the list
        zoneTimes.Clear()
        Wdbg("I", "Cleaned up zoneTimes.")

        'Adds date and time to every single time zone to the list
        For Each zone In zones
            zoneTimes.Add(zone.Id, ConvertTime(KernelDateTime, FindSystemTimeZoneById(zone.Id)))
        Next

    End Sub

    ''' <summary>
    ''' Shows current time in selected time zone, or all of them if zone was "all" 
    ''' </summary>
    ''' <param name="zone">Time zone</param>
    Public Sub ShowTimesInZones(Optional ByVal zone As String = "all")

        If zoneTimes.Keys.Contains(zone) Then
            Write(DoTranslation("- Time of {0}: {1}") + " ({2})", True, ColTypes.Neutral, zone, zoneTimes(zone).ToString(), FindSystemTimeZoneById(zone).GetUtcOffset(KernelDateTime).ToString)
        End If
        If zone = "all" Then
            For Each timezone In zoneTimes.Keys
                Write(DoTranslation("- Time of {0}: {1}") + " ({2})", True, ColTypes.Neutral, timezone, zoneTimes(timezone).ToString(), FindSystemTimeZoneById(timezone).GetUtcOffset(KernelDateTime).ToString)
            Next
        End If

    End Sub

End Module

File: Kernel Simulator/TimeDate/TimeZones.cs
using System;
using System.Collections.Generic;
using System.Linq;

// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using static System.TimeZoneInfo;
using KS.ConsoleBase.Colors;
using KS.Languages;
using KS.Misc.Writers.ConsoleWriters;
using KS.Misc.Writers.DebugWriters;

namespace KS.TimeDate
{
	public static class TimeZones
	{

		/// <summary>
        /// Populates current time in all of the time zones (IANA on Unix).
        /// </summary>
		public static Dictionary<string, DateTime> GetTimeZones()
		{
			// Get all system time zones (IANA on Unix)
			TimeZoneInfo[] Zones = TimeZoneInfo.GetSystemTimeZones().ToArray();
			var ZoneTimes = new Dictionary<string, DateTime>();
			DebugWriter.Wdbg(DebugLevel.I, "Found {0} time zones.", Zones.Length);

			// Run a cleanup in the list
			ZoneTimes.Clear();
			DebugWriter.Wdbg(DebugLevel.I, "Cleaned up zoneTimes.");

			// Adds date and time to every single time zone to the list
			foreach (var Zone in Zones)
				ZoneTimes.Add(Zone.Id, ConvertTime(TimeDate.KernelDateTime, FindSystemTimeZoneById(Zone.Id)));

			// Return the populated array
			return ZoneTimes;
		}

		/// <summary>
        /// Shows current time in selected time zone
        /// </summary>
        /// <param name="zone">Time zone</param>
        /// <returns>True if found; False if not found</returns>
		public static bool ShowTimeZone(string Zone)
		{
			var ZoneTimes = GetTimeZones();
			bool ZoneFound = ZoneTimes.ContainsKey(Zone);
			if (ZoneFound)
			{
				TextWriterColor.Write(Translate.DoTranslation("- Time of {0}: {1}") + " ({2})", true, color: KernelColorTools.GetConsoleColor(KernelColorTools.ColTypes.Neutral), Zone, ZoneTimes[Zone].ToString(), FindSystemTimeZoneById(Zone).GetUtcOffset(TimeDate.KernelDateTime).ToString());
			}
			return ZoneFound;
		}

		/// <summary>
        /// Shows current time in selected time zone
        /// </summary>
        /// <param name="zone">Time zone to search</param>
        /// <returns>True if found; False if not found</returns>
		public static bool ShowTimeZones(string Zone)
		{
			var ZoneTimes = GetTimeZones();
			var ZoneFound = default(bool);
			foreach (string ZoneName in ZoneTimes.Keys)
			{
				if (ZoneName.Contains(Zone))
				{
					ZoneFound = true;
					TextWriterColor.Write(Translate.DoTranslation("- Time of {0}: {1}") + " ({2})", true, color: KernelColorTools.GetConsoleColor(KernelColorTools.ColTypes.Neutral), ZoneName, ZoneTimes[ZoneName].ToString(), FindSystemTimeZoneById(ZoneName).GetUtcOffset(TimeDate.KernelDateTime).ToString());
				}
			}
			return ZoneFound;
		}

		/// <summary>
        /// Shows current time in all time zones
        /// </summary>
		public static void ShowAllTimeZones()
		{
			var ZoneTimes = GetTimeZones();
			foreach (var TimeZone in ZoneTimes.Keys)
				TextWriterColor.Write(Translate.DoTranslation("- Time of {0}: {1}") + " ({2})", true, color: KernelColorTools.GetConsoleColor(KernelColorTools.ColTypes.Neutral), TimeZone, ZoneTimes[TimeZone].ToString(), FindSystemTimeZoneById(TimeZone).GetUtcOffset(TimeDate.KernelDateTime).ToString());
		}

	}
}